File: tests/bindings/java/MetadataResolverFixture.java
package com.statespec.generated;

import com.statespec.backend.Backend;
import com.statespec.backend.ExternalSystem;
import com.statespec.backend.Json;
import com.statespec.generated.external.metadata.*;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public final class MetadataResolverFixture
{
    private static final class FixtureTx implements Backend.Transaction
    {
        @Override public boolean isOpen()
        {
            return true;
        }
        @Override public void abort() throws Backend.BackendException {}

        @Override
        public Optional<Backend.VersionedRecord>
        get(String collection,
            String key) throws Backend.BackendException
        {
            throw new Backend.BackendException("unsupported fixture transaction get");
        }

        @Override
        public List<Backend.VersionedRecord> query(
            String collection,
            Backend.Query query
        ) throws Backend.BackendException
        {
            throw new Backend.BackendException("unsupported fixture transaction query");
        }

        @Override
        public void
        put(String collection,
            String key,
            Json document) throws Backend.BackendException
        {
            throw new Backend.BackendException("unsupported fixture transaction put");
        }

        @Override
        public void erase(
            String collection,
            String key
        ) throws Backend.BackendException
        {
            throw new Backend.BackendException("unsupported fixture transaction erase");
        }
    }

    private static final class FixtureResolver implements ExternalSystem
    {
        int calls = 0;

        @Override
        public Optional<MetadataResolution> resolveMetadata(
            Backend backend,
            MetadataLookup lookup
        ) throws Backend.BackendException
        {
            return Optional.empty();
        }

        @Override
        public Optional<MetadataResolution> resolveMetadataTx(
            Backend.Transaction tx,
            MetadataLookup lookup
        ) throws Backend.BackendException
        {
            calls++;
            Json document = Json.object(Map.of(
                "tenant_id", Json.string("tenant-a"), "base_url",
                Json.string("https://api.stripe.test")
            ));
            return Optional.of(new MetadataResolution(
                new Backend.VersionedRecord(
                    lookup.metadataEntity(), "tenant-a/stripe/default", 1L, document
                ),
                ExternalSystem.missingRequiredMetadataFields(document, lookup.requiredFields())
            ));
        }
    }

    private static final class FixtureMappingApplicator
        implements ExternalSystemMetadataMappingApplicator
    {
        @Override
        public ExternalSystemMetadataMappingOutput applyExternalSystemMetadataMappings(
            ExternalSystemMetadataMappingPlan plan,
            ExternalSystemMetadataMappingInputs inputs
        )
        {
            java.util.HashMap<String, Json> clientConfig = new java.util.HashMap<>();
            java.util.HashMap<String, Json> requestPayload = new java.util.HashMap<>();
            for (ExternalSystemMetadataMappingAssignment assignment :
                 plan.allMappings())
            {
                Optional<Json> value = inputs.assignmentValue(assignment);
                if (value.isEmpty())
                {
                    continue;
                }
                if (assignment.targetRoot().equals("client"))
                {
                    clientConfig.put(assignment.field(), value.orElseThrow());
                }
                else if (assignment.targetRoot().equals("request"))
                {
                    requestPayload.put(assignment.field(), value.orElseThrow());
                }
            }
            return new ExternalSystemMetadataMappingOutput(
                clientConfig, requestPayload,
                ExternalSystemMetadata.missingExternalSystemMetadataMappingSources(plan, inputs)
            );
        }
    }

    private static final class FixtureOperatorMetadataRepository
        implements ExternalSystemOperatorMetadataRepository
    {
        @Override
        public Optional<Backend.VersionedRecord> upsertMetadataTx(
            Backend.Transaction tx,
            ExternalSystemOperatorMetadataUpsertRequest request
        )
        {
            return Optional.of(new Backend.VersionedRecord(
                request.lookup().metadataEntity(), "tenant-a/stripe/default",
                request.expectedVersion().orElse(0L) + 1L, request.document()
            ));
        }

        @Override
        public Optional<Backend.VersionedRecord> getMetadataTx(
            Backend.Transaction tx,
            ExternalSystemOperatorMetadataGetRequest request
        )
        {
            return Optional.of(new Backend.VersionedRecord(
                request.lookup().metadataEntity(), "tenant-a/stripe/default", 1L,
                Json.object(Map.of("status", Json.string("Active")))
            ));
        }

        @Override
        public Optional<Backend.VersionedRecord> disableMetadataTx(
            Backend.Transaction tx,
            ExternalSystemOperatorMetadataDisableRequest request
        )
        {
            return Optional.of(new Backend.VersionedRecord(
                request.lookup().metadataEntity(), "tenant-a/stripe/default",
                request.expectedVersion().orElse(0L) + 1L,
                Json.object(Map.of("status", Json.string(request.disabledStatus())))
            ));
        }

        @Override
        public Optional<Backend.VersionedRecord> deleteMetadataTx(
            Backend.Transaction tx,
            ExternalSystemOperatorMetadataDeleteRequest request
        )
        {
            return Optional.of(new Backend.VersionedRecord(
                request.lookup().metadataEntity(), "tenant-a/stripe/default",
                request.expectedVersion().orElse(0L) + 1L,
                Json.object(Map.of("status", Json.string(request.deletedStatus())))
            ));
        }
    }

    private static final class FixtureOperatorMetadataApiHandler
        implements ExternalSystemOperatorMetadataApiHandler
    {
        @Override
        public Descriptors.ApiResponse handleUpsertMetadataTx(
            Backend.Transaction tx,
            ExternalSystemOperatorMetadataRepository repository,
            ExternalSystemOperatorMetadataUpsertRequest request
        ) throws Exception
        {
            Optional<Backend.VersionedRecord> record = repository.upsertMetadataTx(tx, request);
            return new Descriptors.ApiResponse(
                record.isPresent() ? 200 : 404,
                Json.object(Map.of("operation", Json.string("upsert")))
            );
        }

        @Override
        public Descriptors.ApiResponse handleGetMetadataTx(
            Backend.Transaction tx,
            ExternalSystemOperatorMetadataRepository repository,
            ExternalSystemOperatorMetadataGetRequest request
        ) throws Exception
        {
            Optional<Backend.VersionedRecord> record = repository.getMetadataTx(tx, request);
            return new Descriptors.ApiResponse(
                record.isPresent() ? 200 : 404, Json.object(Map.of("operation", Json.string("get")))
            );
        }

        @Override
        public Descriptors.ApiResponse handleDisableMetadataTx(
            Backend.Transaction tx,
            ExternalSystemOperatorMetadataRepository repository,
            ExternalSystemOperatorMetadataDisableRequest request
        ) throws Exception
        {
            Optional<Backend.VersionedRecord> record = repository.disableMetadataTx(tx, request);
            return new Descriptors.ApiResponse(
                record.isPresent() ? 200 : 404,
                Json.object(Map.of("operation", Json.string("disable")))
            );
        }

        @Override
        public Descriptors.ApiResponse handleDeleteMetadataTx(
            Backend.Transaction tx,
            ExternalSystemOperatorMetadataRepository repository,
            ExternalSystemOperatorMetadataDeleteRequest request
        ) throws Exception
        {
            Optional<Backend.VersionedRecord> record = repository.deleteMetadataTx(tx, request);
            return new Descriptors.ApiResponse(
                record.isPresent() ? 200 : 404,
                Json.object(Map.of("operation", Json.string("delete")))
            );
        }
    }

    public static void main(String[] args) throws Exception
    {
        FixtureResolver resolver = new FixtureResolver();
        FixtureTx tx = new FixtureTx();
        ExternalSystemMetadataMappingPlan plan =
            ExternalSystemMetadata.externalSystemMetadataMappingPlan(
                Descriptors.externalSystemDescriptors().get(0)
            );
        if (plan.allMappings().size() != 4 || plan.clientMappings().size() != 3 ||
            plan.requestMappings().size() != 1 ||
            !plan.clientMappings().get(0).sourceRoot().equals("metadata") ||
            !plan.clientMappings().get(0).sourceField().equals("base_url") ||
            !plan.clientMappings().get(0).targetRoot().equals("client") ||
            !plan.clientMappings().get(0).field().equals("base_url") ||
            !plan.requestMappings().get(0).sourceRoot().equals("input") ||
            !plan.requestMappings().get(0).sourceField().equals("order_id") ||
            !plan.requestMappings().get(0).targetRoot().equals("request") ||
            !plan.requestMappings().get(0).field().equals("order_id"))
        {
            throw new AssertionError("unexpected metadata mapping plan");
        }
        ExternalSystemMetadataMappingApplicator applicator =
            new FixtureMappingApplicator();
        ExternalSystemMetadataMappingOutput mapped =
            applicator.applyExternalSystemMetadataMappings(
                plan, new ExternalSystemMetadataMappingInputs(
                          Map.of("order_id", Json.string("order-1")), Map.of(), Map.of(),
                          Map.of(
                              "base_url", Json.string("https://api.stripe.test"), "auth_ref",
                              Json.string("secret:stripe"), "timeout_ms", Json.integer(5000)
                          )
                      )
            );
        if (mapped.clientConfig().size() != 3 || mapped.requestPayload().size() != 1 ||
            !mapped.missingSources().isEmpty())
        {
            throw new AssertionError("unexpected mapped metadata output");
        }
        ExternalSystemMetadataMappingOutput missingMapped =
            applicator.applyExternalSystemMetadataMappings(
                plan, new ExternalSystemMetadataMappingInputs(
                          Map.of("order_id", Json.string("order-1")), Map.of(), Map.of(),
                          Map.of(
                              "base_url", Json.string("https://api.stripe.test"), "auth_ref",
                              Json.string("secret:stripe")
                          )
                      )
            );
        if (missingMapped.missingSources().size() != 1 ||
            !missingMapped.missingSources().get(0).source().equals("metadata.timeout_ms") ||
            !missingMapped.missingSources().get(0).targetRoot().equals("client") ||
            !missingMapped.missingSources().get(0).field().equals("timeout_ms"))
        {
            throw new AssertionError("unexpected missing mapping diagnostics");
        }
        List<ExternalSystem.MetadataKeyValue> keys = List.of(
            new ExternalSystem.MetadataKeyValue("tenant_id", Json.string("tenant-a")),
            new ExternalSystem.MetadataKeyValue("external_system_id", Json.string("stripe")),
            new ExternalSystem.MetadataKeyValue("profile", Json.string("default"))
        );
        ExternalSystem.MetadataLookup lookup =
            ExternalSystemMetadata.externalSystemMetadataLookup("Stripe", keys).orElseThrow();
        ExternalSystemOperatorMetadataRepository repository =
            new FixtureOperatorMetadataRepository();
        Backend.VersionedRecord upserted =
            repository
                .upsertMetadataTx(
                    tx, new ExternalSystemOperatorMetadataUpsertRequest(
                            lookup, Json.object(Map.of("tenant_id", Json.string("tenant-a"))),
                            Optional.of(1L)
                        )
                )
                .orElseThrow();
        Backend.VersionedRecord loaded =
            repository
                .getMetadataTx(tx, new ExternalSystemOperatorMetadataGetRequest(lookup))
                .orElseThrow();
        Backend.VersionedRecord disabled =
            repository
                .disableMetadataTx(
                    tx, new ExternalSystemOperatorMetadataDisableRequest(
                            lookup, Optional.of(upserted.version()), "Disabled"
                        )
                )
                .orElseThrow();
        Backend.VersionedRecord deleted =
            repository
                .deleteMetadataTx(
                    tx, new ExternalSystemOperatorMetadataDeleteRequest(
                            lookup, Optional.of(disabled.version()), "Deleted"
                        )
                )
                .orElseThrow();
        if (upserted.version() != 2L || loaded.version() != 1L || disabled.version() != 3L ||
            deleted.version() != 4L)
        {
            throw new AssertionError("unexpected metadata repository result");
        }
        ExternalSystemOperatorMetadataApiHandler metadataApiHandler =
            new FixtureOperatorMetadataApiHandler();
        Descriptors.ApiResponse metadataApiResponse = metadataApiHandler.handleUpsertMetadataTx(
            tx, repository,
            new ExternalSystemOperatorMetadataUpsertRequest(
                lookup, Json.object(Map.of("tenant_id", Json.string("tenant-a"))),
                Optional.of(deleted.version())
            )
        );
        if (metadataApiResponse.statusCode() != 200)
        {
            throw new AssertionError("unexpected operator metadata API response");
        }
        Optional<ExternalSystem.MetadataResolution> resolved =
            ExternalSystemMetadata.resolveExternalSystemMetadataTx(resolver, tx, "Stripe", keys);
        if (resolved.isEmpty() || resolved.orElseThrow().complete() || resolver.calls != 1)
        {
            throw new AssertionError("expected incomplete metadata resolution through resolver");
        }

        Optional<ExternalSystem.MetadataResolution> skipped =
            ExternalSystemMetadata.resolveExternalSystemMetadataTx(resolver, tx, "Stripe", keys.subList(0, 2));
        if (skipped.isPresent() || resolver.calls != 1)
        {
            throw new AssertionError("expected incomplete key to skip resolver");
        }
    }
}

File: tests/bindings/cpp/metadata_resolver_fixture.cpp
#include "common/system_descriptors.hpp"

#include <string>
#include <vector>

class FakeTx final : public statespec::backend::ITransaction
{
  public:
    bool is_open() const override
    {
        return true;
    }
    void abort() override {}
};

class FakeResolver final : public statespec::backend::IExternalSystemMetadataResolver
{
  public:
    int calls = 0;

    std::optional<statespec::backend::ExternalSystemMetadataResolution> resolve_metadata(
        statespec::backend::IBackend&,
        const statespec::backend::ExternalSystemMetadataLookup&
    ) override
    {
        return std::nullopt;
    }

    std::optional<statespec::backend::ExternalSystemMetadataResolution> resolve_metadataTx(
        statespec::backend::ITransaction&,
        const statespec::backend::ExternalSystemMetadataLookup& lookup
    ) override
    {
        ++calls;
        const auto document = statespec::backend::Json::object(
            {{"tenant_id", "tenant-a"}, {"base_url", "https://api.stripe.test"}}
        );
        return statespec::backend::ExternalSystemMetadataResolution{
            statespec::backend::VersionedRecord{
                lookup.metadata_entity,
                "tenant-a/stripe/default",
                1,
                document,
            },
            statespec::backend::missing_required_metadata_fields(document, lookup.required_fields),
        };
    }
};

class FakeMappingApplicator final
    : public statespec_generated::IExternalSystemMetadataMappingApplicator
{
  public:
    statespec_generated::ExternalSystemMetadataMappingOutput apply(
        const statespec_generated::ExternalSystemMetadataMappingPlan& plan,
        const statespec_generated::ExternalSystemMetadataMappingInputs& inputs
    ) override
    {
        statespec_generated::ExternalSystemMetadataMappingOutput output;
        output.missing_sources =
            statespec_generated::missing_external_system_metadata_mapping_sources(plan, inputs);
        for (const auto& assignment : plan.all_mappings)
        {
            const auto* source = inputs.source_value(assignment);
            if (source == nullptr)
            {
                continue;
            }
            if (assignment.target_root == "client")
            {
                output.client_config.emplace(assignment.field, *source);
            }
            else if (assignment.target_root == "request")
            {
                output.request_payload.emplace(assignment.field, *source);
            }
        }
        return output;
    }
};

class FakeOperatorMetadataRepository final
    : public statespec_generated::IExternalSystemOperatorMetadataRepository
{
  public:
    std::optional<statespec::backend::VersionedRecord> upsert_metadataTx(
        statespec::backend::ITransaction&,
        const statespec_generated::ExternalSystemOperatorMetadataUpsertRequest& request
    ) override
    {
        return statespec::backend::VersionedRecord{
            request.lookup.metadata_entity,
            "tenant-a/stripe/default",
            request.expected_version.value_or(0) + 1,
            request.document,
        };
    }

    std::optional<statespec::backend::VersionedRecord> get_metadataTx(
        statespec::backend::ITransaction&,
        const statespec_generated::ExternalSystemOperatorMetadataGetRequest& request
    ) override
    {
        return statespec::backend::VersionedRecord{
            request.lookup.metadata_entity,
            "tenant-a/stripe/default",
            1,
            statespec::backend::Json::object({{"status", "Active"}}),
        };
    }

    std::optional<statespec::backend::VersionedRecord> disable_metadataTx(
        statespec::backend::ITransaction&,
        const statespec_generated::ExternalSystemOperatorMetadataDisableRequest& request
    ) override
    {
        return statespec::backend::VersionedRecord{
            request.lookup.metadata_entity,
            "tenant-a/stripe/default",
            request.expected_version.value_or(0) + 1,
            statespec::backend::Json::object({{"status", request.disabled_status}}),
        };
    }

    std::optional<statespec::backend::VersionedRecord> delete_metadataTx(
        statespec::backend::ITransaction&,
        const statespec_generated::ExternalSystemOperatorMetadataDeleteRequest& request
    ) override
    {
        return statespec::backend::VersionedRecord{
            request.lookup.metadata_entity,
            "tenant-a/stripe/default",
            request.expected_version.value_or(0) + 1,
            statespec::backend::Json::object({{"status", request.deleted_status}}),
        };
    }
};

class FakeOperatorMetadataApiHandler final
    : public statespec_generated::IExternalSystemOperatorMetadataApiHandler
{
  public:
    statespec_generated::ApiResponse handle_upsert_metadataTx(
        statespec::backend::ITransaction& tx,
        statespec_generated::IExternalSystemOperatorMetadataRepository& repository,
        const statespec_generated::ExternalSystemOperatorMetadataUpsertRequest& request
    ) override
    {
        auto record = repository.upsert_metadataTx(tx, request);
        return statespec_generated::ApiResponse{
            record.has_value() ? 200 : 404,
            statespec::backend::Json::object({{"operation", "upsert"}}),
        };
    }

    statespec_generated::ApiResponse handle_get_metadataTx(
        statespec::backend::ITransaction& tx,
        statespec_generated::IExternalSystemOperatorMetadataRepository& repository,
        const statespec_generated::ExternalSystemOperatorMetadataGetRequest& request
    ) override
    {
        auto record = repository.get_metadataTx(tx, request);
        return statespec_generated::ApiResponse{
            record.has_value() ? 200 : 404,
            statespec::backend::Json::object({{"operation", "get"}}),
        };
    }

    statespec_generated::ApiResponse handle_disable_metadataTx(
        statespec::backend::ITransaction& tx,
        statespec_generated::IExternalSystemOperatorMetadataRepository& repository,
        const statespec_generated::ExternalSystemOperatorMetadataDisableRequest& request
    ) override
    {
        auto record = repository.disable_metadataTx(tx, request);
        return statespec_generated::ApiResponse{
            record.has_value() ? 200 : 404,
            statespec::backend::Json::object({{"operation", "disable"}}),
        };
    }

    statespec_generated::ApiResponse handle_delete_metadataTx(
        statespec::backend::ITransaction& tx,
        statespec_generated::IExternalSystemOperatorMetadataRepository& repository,
        const statespec_generated::ExternalSystemOperatorMetadataDeleteRequest& request
    ) override
    {
        auto record = repository.delete_metadataTx(tx, request);
        return statespec_generated::ApiResponse{
            record.has_value() ? 200 : 404,
            statespec::backend::Json::object({{"operation", "delete"}}),
        };
    }
};

int main()
{
    FakeTx tx;
    FakeResolver resolver;
    const auto descriptors = statespec_generated::external_system_descriptors();
    const auto plan = statespec_generated::external_system_metadata_mapping_plan(descriptors[0]);
    if (plan.all_mappings.size() != 4 || plan.client_mappings.size() != 3 ||
        plan.request_mappings.size() != 1 || plan.client_mappings[0].source_root != "metadata" ||
        plan.client_mappings[0].source_field != "base_url" ||
        plan.client_mappings[0].target_root != "client" ||
        plan.client_mappings[0].field != "base_url" ||
        plan.request_mappings[0].source_root != "input" ||
        plan.request_mappings[0].source_field != "order_id" ||
        plan.request_mappings[0].target_root != "request" ||
        plan.request_mappings[0].field != "order_id")
    {
        return 1;
    }
    FakeMappingApplicator applicator;
    statespec_generated::IExternalSystemMetadataMappingApplicator& mapping_applicator = applicator;
    const auto mapped = mapping_applicator.apply(
        plan, statespec_generated::ExternalSystemMetadataMappingInputs{
                  {{"order_id", "order-1"}},
                  {},
                  {},
                  {{"base_url", "https://api.stripe.test"},
                   {"auth_ref", "secret:stripe"},
                   {"timeout_ms", std::int64_t{5000}}},
              }
    );
    if (mapped.client_config.size() != 3 || mapped.request_payload.size() != 1 ||
        !mapped.missing_sources.empty() ||
        mapped.client_config.find("base_url") == mapped.client_config.end() ||
        mapped.request_payload.find("order_id") == mapped.request_payload.end())
    {
        return 1;
    }
    const auto missing_mapped = mapping_applicator.apply(
        plan, statespec_generated::ExternalSystemMetadataMappingInputs{
                  {{"order_id", "order-1"}},
                  {},
                  {},
                  {{"base_url", "https://api.stripe.test"}, {"auth_ref", "secret:stripe"}},
              }
    );
    if (missing_mapped.missing_sources.size() != 1 ||
        missing_mapped.missing_sources[0].source != "metadata.timeout_ms" ||
        missing_mapped.missing_sources[0].target_root != "client" ||
        missing_mapped.missing_sources[0].field != "timeout_ms")
    {
        return 1;
    }
    std::vector<statespec::backend::ExternalSystemMetadataKeyValue> keys{
        {"tenant_id", "tenant-a"},
        {"external_system_id", "stripe"},
        {"profile", "default"},
    };
    auto lookup = statespec_generated::external_system_metadata_lookup("Stripe", keys);
    if (!lookup.has_value())
    {
        return 1;
    }
    FakeOperatorMetadataRepository repository;
    statespec_generated::IExternalSystemOperatorMetadataRepository& metadata_repository =
        repository;
    const auto upserted = metadata_repository.upsert_metadataTx(
        tx, statespec_generated::ExternalSystemOperatorMetadataUpsertRequest{
                *lookup,
                statespec::backend::Json::object({{"tenant_id", "tenant-a"}}),
                statespec::backend::Version{1},
            }
    );
    const auto loaded = metadata_repository.get_metadataTx(
        tx, statespec_generated::ExternalSystemOperatorMetadataGetRequest{*lookup}
    );
    const auto disabled = metadata_repository.disable_metadataTx(
        tx, statespec_generated::ExternalSystemOperatorMetadataDisableRequest{
                *lookup,
                statespec::backend::Version{2},
                "Disabled",
            }
    );
    const auto deleted = metadata_repository.delete_metadataTx(
        tx, statespec_generated::ExternalSystemOperatorMetadataDeleteRequest{
                *lookup,
                statespec::backend::Version{3},
                "Deleted",
            }
    );
    if (!upserted.has_value() || upserted->version != 2 || !loaded.has_value() ||
        !disabled.has_value() || disabled->version != 3 || !deleted.has_value() ||
        deleted->version != 4)
    {
        return 1;
    }
    FakeOperatorMetadataApiHandler operator_api_handler;
    statespec_generated::IExternalSystemOperatorMetadataApiHandler& metadata_api_handler =
        operator_api_handler;
    const auto api_response = metadata_api_handler.handle_upsert_metadataTx(
        tx, metadata_repository,
        statespec_generated::ExternalSystemOperatorMetadataUpsertRequest{
            *lookup,
            statespec::backend::Json::object({{"tenant_id", "tenant-a"}}),
            statespec::backend::Version{4},
        }
    );
    if (api_response.status_code != 200)
    {
        return 1;
    }
    auto resolved =
        statespec_generated::resolve_external_system_metadataTx(resolver, tx, "Stripe", keys);
    if (!resolved.has_value() || resolved->complete() || resolver.calls != 1)
    {
        return 1;
    }

    keys.pop_back();
    auto skipped =
        statespec_generated::resolve_external_system_metadataTx(resolver, tx, "Stripe", keys);
    return skipped.has_value() || resolver.calls != 1 ? 1 : 0;
}
